### ISSUE TITLE
IDE cannot resolve module tf.keras

### ISSUE THREAD

**Issue by ctrysbita** [tensorflow/tensorflow#26502]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
No.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Linux Ubuntu 18.10
- TensorFlow installed from (source or binary):
binary (pip)
- TensorFlow version (use command below):
2.0_alpha
- Python version:
3.6

**Describe the current behavior**
`import tensorflow as tf`
then pyCharm cannot resolve module tf.keras and report an error:
`Cannot find reference 'keras' in '__init__.py'`
But when running program, everything works well.

**Describe the expected behavior**
tf.keras imported successfully with autocomplete of pyCharm.

**Code to reproduce the issue**
`import tensorflow as tf`

**Other info / logs**
It seems that there is no import command for `keras` module in `__init_.py` of tensorflow package.
When I added `from tensorflow.python import keras` to `__init__.py`manually, everything work well.
Maybe there are some problem for package importing after `keras` was moved from `_api` to `python`.

**Comment by knaydenov** [comment]
https://stackoverflow.com/a/47306203/6108843

**Comment by jaingaurav** [comment]
Not very familiar with how these modules work. @annarev would you know what's up?

**Comment by knaydenov** [comment]
https://intellij-support.jetbrains.com/hc/en-us/community/posts/360002486739/comments/360000407199

**Comment by ErmiaAzarkhalili** [comment]
@knaydenov none of your solutions is pythonic. It should work like @tensorflow import keras in its examples.

**Comment by ErmiaAzarkhalili** [comment]
@annarev any solution to this issue? any commit for the next dev or stable release?

**Comment by ctrysbita** [comment]
Yes, this is not a bug from pyCharm but tensorflow itself.
The bug is caused by missing `tensorflow.python import keras` in `__init__.py` of tensorflow package.

**Comment by ErmiaAzarkhalili** [comment]
@ctrysbita do you think they resolved the issue in future releases?

**Comment by annarev** [comment]
keras is imported dynamically in the __init__.py file. 
I am not familiar with the way pyCharm looks for autocomplete symbols, but I will see if there is any workaround.

**Comment by ErmiaAzarkhalili** [comment]
> keras is imported dynamically in the `__init__.py` file.
> I am not familiar with the way pyCharm looks for autocomplete symbols, but I will see if there is any workaround (may be adding 'keras' to `__all__`).

As we discussed here, It's a serious problem which is created by tensorflow implementation. Could you please kindly consider it to resolve as soon as possible for next versions?

**Comment by yzho0907** [comment]
seems this issue hasn't been resolved yet, any idea to fix it temporarily? thx!

**Comment by ihexx** [comment]
> seems this issue hasn't been resolved yet, any idea to fix it temporarily? thx!

At least for code completion to find it, a quick and dirty fix is to throw `from tensorflow.python import keras` at the top of <python_site_packages_dir>/tensorflow/__init__.py

**Comment by ErmiaAzarkhalili** [comment]
@jaingaurav Hello, could you please let us when this bug will be fixed?

**Comment by ma3oun** [comment]
> At least for code completion to find it, a quick and dirty fix is to throw `from tensorflow.python import keras` at the top of `<python_site_packages_dir>/tensorflow/__init__.py`
> In pycharm you can get there quickly by just going to the declaration of tensorflow (ctrl+b default)

Doing this solved the cde completion issue but led to other errors, namely API incompatibilities. I tried creating my own Model by subclassing and I got strange errors from the _tensorflow_backend.py_ functions.

**Comment by jaingaurav** [comment]
For folks hitting this issue, the fix has merged. Please try the latest 2.0 nightly and let us know if it is not yet resolved.

**Comment by DecentGradient** [comment]
Not resolved for me in pycharm

**Comment by annarev** [comment]
@DecentGradient which version of TensorFlow are you using?

**Comment by DecentGradient** [comment]
2.0 nightly updated today

**Comment by DecentGradient** [comment]
CPU

**Comment by DecentGradient** [comment]
@annarev tf-nightly-2.0-preview-2.0.0.dev20190430

**Comment by annarev** [comment]
I just installed latest tf-nightly-2.0-preview and can't repro the issue. For example, I see autocompletion suggestions in PyCharms when typing `tf.keras.applications`.

Can you describe the issue you are seeing in more detail? Which name doesn't autocomplete?

**Comment by DecentGradient** [comment]
@annarev when i import tensorflow.keras the ide says `Cannot find reference 'keras' in '__init__.py' ` and cannot autocomplete anything under keras

**Comment by DecentGradient** [comment]
Interesting , @annarev 
when using ` import tensorflow as tf ` then `tf.keras.*` auto completes 
But  `from tensorflow.keras import layers` will still not "work"

**Comment by mihaimaruseac** [comment]
There is a difference between the two: `tf.keras.*` uses the attribute `keras` of module imported as `tf` but `import tensorflow.keras` looks for module `keras` in package `tensorflow`.

Now we know where to look

**Comment by ErmiaAzarkhalili** [comment]
@annarev when using `import tensorflow as tf` then tf.keras.* autocompletes
But from tensorflow.keras import layers will still not "work"!

**Comment by ErmiaAzarkhalili** [comment]
> There is a difference between the two: `tf.keras.*` uses the attribute `keras` of module imported as `tf` but `import tensorflow.keras` looks for module `keras` in package `tensorflow`.

It doesn't make sense to me. 
why `import tensorflow.keras` doesn't work properly?

**Comment by ErmiaAzarkhalili** [comment]
> > There is a difference between the two: `tf.keras.*` uses the attribute `keras` of module imported as `tf` but `import tensorflow.keras` looks for module `keras` in package `tensorflow`.
> 
> It doesn't make sense to me.
> why `import tensorflow.keras` doesn't work properly?
I would like

**Comment by annarev** [comment]
I will reopen the bug. I suspect getting `import tensorflow.keras` would be much harder to fix. Especially, if it depends on having `keras` under `tensorflow` directory (which is not possible since these are installed as separate pip packages and live in different directories).

**Comment by phdsky** [comment]
Problem still existing in the lastest tensorflow-gpu-2.0.0a0, Please fix this.

Insight normal in the terminal, but not in pycharm

**Comment by mihaimaruseac** [comment]
I think the issue comes from the 3 different ways we are doing imports:

* normal Python import (`import foo` and `from foo import bar`)
* imports by extending `sys.path` and `sys.modules` and `__path__`
* using `LazyLoader` to make a name available, loading the corresponding module lazily

I think that the last two are the ones causing issues and this will be hard to fix at the moment. There's another issue caused by the same root cause, just linking it here for completeness of the image: #28214

**Comment by annarev** [comment]
@phdsk tensorflow-gpu-2.0.0a0 was released before the fix for `import tensorflow as tf` went in. That fix is only in nightly. I don't know if you are using `import tensorflow as tf`, but if you are, you can nightly.
If you are using other ways of importing tensorflow, then pycharms suggestions won't work in any version.

**Comment by annarev** [comment]
@ErmiaAzarkhalili to workaround `import tensorflow.keras as tfk`, you can use `from tensorflow import keras as tfk`

**Comment by PhanDuc** [comment]
On Pycharm, instead of `from tensorflow import keras`, try to use` from tensorflow.python import keras`  will help autocomplete. I tested with Tensorlow 2.0 alpha and it is working.

**Comment by ivanlv** [comment]
As said by @PhanDuc, it works for me on Tensorflow 2.0 alpha.

**Comment by jvishnuvardhan** [comment]
Closing this out since I understand it to be resolved, but please let me know if I'm mistaken. Thanks!

**Comment by DecentGradient** [comment]
I believe this is unresolved as the current working suggestion is a temporary workaround

**Comment by annarev** [comment]
I would say the original issue is resolved since it talks about `import tensorflow as tf` specifically.
However, `from tensorflow.keras import losses` and similar are not resolved yet. May be it is better to open a new issue?

**Comment by Burton2000** [comment]
Seems like this is now the issue https://github.com/tensorflow/tensorflow/issues/29144

**Comment by subedigaurav** [comment]
I had the same problem and it bugged me for a couple of hours :(. I used `from tensorflow.python.keras import ***` for importing the libraries and it worked fine for me.

**Comment by ErmiaAzarkhalili** [comment]
> @ErmiaAzarkhalili to workaround `import tensorflow.keras as tfk`, you can use `from tensorflow import keras as tfk`

Thanks, but is it pythonic solution? It seems circumvention to me, not a real solution.

**Comment by PhanDuc** [comment]
Well, for loss functions ` from tensorflow.python.keras.losses import *` or simple keras `from tensorflow.python import keras` should work.

**Comment by OmriPi** [comment]
This occurs also in 1.14, and renders PyCharm completely useless with TensorFlow. It used to be OK, what happened?

**Comment by bencbartlett** [comment]
I am also still having this issue running PyCharm 2019.2 with Python 3.7 and `tf-2.0.0-beta1`. If I run `from tensorflow import keras` in a Python interpreter it works fine, but tf seems to have done something that renders PyCharm unable to find many of its submodules, which makes it difficult to work with due to the lack of autocomplete and false import warnings.

**Comment by yiyang-yu** [comment]
> I had the same problem and it bugged me for a good couple of hours :( . I used `from tensorflow.python.keras import ***` for importing the libraries and it worked fine for me.

Thanks for the temporary solution! It had been working for a while; but now with `2.0.0-dev20190805`, even this solution does not work: no way to make PyCharm(2019.2) find the correct reference inside TensorFlow.

**Comment by procule** [comment]
Using `from tensorflow import keras` works but since almost all documentations / examples / books for TF 2.0 are using `tf.keras`, that's still a problem.

An bug tracking ticket for PyCharm is opened here: https://youtrack.jetbrains.com/issue/PY-34174

**Comment by LaCandela** [comment]
I am still having this issue with pycharm 2019.1.3
I cannot import keras from tensorflow. "from tensorflow import keras" doesn't work either.
I am using the nightly build of TF2.0: tf-nightly-gpu-2.0-preview 2.0.0.dev20190821

**Comment by georgealexandruvlad** [comment]
Same issue with the new tensorflow==2.0.0-rc0 release. If in beta release I could just use the tensorflow.python.keras workaround, now keras.layers module doesn't see the layers classes (Dense, Embedding, etc) but packages in which every layer is grouped. If I import them separately from each module the call of a layer over an input doesn't seem to work. This behaviour might be only a problem in PyCharm (in cmd works just fine) but as the majority of developers are using PyCharm as IDE it would be nice to not struggle to use tensorflow library. In addition the tensorflow namespace doesn't contain anything at import, all the packages are grouped in tensorflow_core. I understand that the backend grouping of modules must be done that way, the problem is that the importing step doesn't match the guides and tutorials on the official site and makes everything hard to use.

If it is any work around on this issue until a fix is released (from tensorflow team or PyCharm) I would be grateful to use it.

**Comment by hankcs** [comment]
Someone should reopen this since neither `tensorflow-gpu-2.0.0rc0` nor `tf-nightly-gpu-2.0-preview` enables IDE to resolve `tf.keras`. I don't really understand why tf2 still uses such a complex importing method. tf1 has reinvented logging and argparse, now tf2 is still reinventing import.

**Comment by SpikingNeuron** [comment]
🤕 I just upgraded to 2.0rc ..... and my code is full of yellows
Earlier for 2.0b I used to use the hack for keras and other stuff
```
import tensorflow.python as tfpy
tfpy.keras
```
Now that doesn't work .... but nonetheless code runs ....

**Comment by deepgradient** [comment]
> I added an update on this issue on the freshly opened #31973 : [#31973 (comment)](https://github.com/tensorflow/tensorflow/issues/31973#issuecomment-524928419)
> 
> TL;DR: we are working on this but the fix is quite complex

Thanks for your great comment. We seriously need this feature; otherwise, pycharm will be completely useless!!!

**Comment by shianiawhite** [comment]
Out of curiosity, what IDEs do the core developers use? I think most heavy-duty IDEs for software development would be affected by this, but I also would have assumed the core developers would use such IDEs. Seeing that this was left in, I'm guessing the developers are using an IDE that wasn't affected.

**Comment by ElPapi42** [comment]
what is going on here? i suffering this error from moths to now

**Comment by TimbusCalin** [comment]
Does not work with the stable version of TensorFlow.Keras!!!!!

**Comment by TimbusCalin** [comment]
Hey everyone, if you still struggle, please check my stackoverflow post in which I solve this issue with
TensorFlow 2.0.0 stable release version.

https://stackoverflow.com/questions/58188704/tensorflow-2-0-unable-to-import-keras-in-pycharm

**Comment by SpikingNeuron** [comment]
At least with PyCharm IDE (EAP release), it will work. I have tried it and it works. [Check here](https://stackoverflow.com/a/58192020/5681083). Avoid importing `tensorflow_core` if you want to keep code compatible.

**Comment by SummitKwan** [comment]
> I had the same problem and it bugged me for a good couple of hours :( . I used `from tensorflow.python.keras import ***` for importing the libraries and it worked fine for me.

That is kind of dangerous since you may end up importing different modules.

e.g.

`from tensorflow.python.keras.callbacks import TensorBoard; print(TensorBoard)` gets
`<class 'tensorflow.python.keras.callbacks.TensorBoard'>`
but, `from tensorflow.keras.callbacks import TensorBoard; print(TensorBoard)` gets `<class 'tensorflow.python.keras.callbacks_v1.TensorBoard'>`

**Comment by ElPapi42** [comment]
This problem is annoying, i end up coding on Google Colab, there the module is recognized and the autocompletion works good (On experimental Colab Gui). I will stay this way while this is solved, but its a pain.

**Comment by enricoros** [comment]
With PyCharm 2019.3 (Early Access Preview or later) the issue disappears.

**Comment by raofengyun** [comment]
why is this issue closed while it's still not fixed?

**Comment by TimbusCalin** [comment]
Apparently they closed the issue because with "PyCharm 2019.3 (Early Access Preview" the issue disappears. However, at the time speaking, PyCharm 2019.3 still has not been released (it is R.C. now)

**Comment by fzyzcjy** [comment]
Same issue with stable Intellij IDEA and tf2 here...

**Comment by annarev** [comment]
@fzyzcjy which import doesn't work for you? Can you try tensorflow 2.2.0rc0 ?

**Comment by fzyzcjy** [comment]
@annarev All does not work... I will first try to update to IDEA 2019.3.4, and then report the situation to you. Thanks!

**Comment by annarev** [comment]
Sounds good. Either updating Intellij IDEA to newer version or updating TensorFlow to 2.2.0rc0 should help. Let me know if you still see issues.

**Comment by fzyzcjy** [comment]
@annarev After updating IDEA to 2019.3.4, it works! Thanks! :)

**Comment by igo312** [comment]
tensorflow-version:'2.0.0' 
CUDA:10.0 Cudnn:7.6.4
(I have tried update` tensorflow` to `2.2.0rc0` but I get `import error:Failed to load the native TensorFlow runtime` ) 

pycharm version: 2019.2.4 -> 2019.3.4

Problem solved!
thanks @annarev and @fzyzcjy

**Comment by elisim** [comment]
tensorflow 1.13.1
PyCharm 2019.3.4
![image](https://user-images.githubusercontent.com/17675462/77259152-40660000-6c88-11ea-80ba-4260dffafced.png)
![image](https://user-images.githubusercontent.com/17675462/77259157-44921d80-6c88-11ea-98fa-0b29b8c1a969.png)



same problem...

**Comment by ctrysbita** [comment]
@elisim I'm using TF2.0 + PyCharm 19.3. Your code works for me.
May be update to latest version can resolve the issue?

**Comment by elisim** [comment]
@ctrysbita It's works with TF2.0, but I need TF1.3 for my needs.

**Comment by darien-schettler** [comment]
This is still occuring with the latest Tensorflow Version installed and the latest version of PyCharm community installed.

Please address this as it's annoying.

**Comment by mihaimaruseac** [comment]
1.3 is too old, we no longer support it or bring fixes to it. Please update to 1.15 or 2.1 (long term releases)

**Comment by lmoesch** [comment]
still not solved for TF 2.1 (Pylance 2020.9.6)

**Comment by zahra-soltaninejad** [comment]
i had such problem with tensorflow keras.leyer dn it doesnt work 
the cod was  from keras.layer import DENSE nd i solve it by running this command in pycharm command erea 

"pip install tensorflow"
